---
title: "Analysis examples in curatedPCaData"
output: 
  rmarkdown::html_vignette:
    toc: true
    number_sections: false
vignette: >
  %\VignetteIndexEntry{Analysis examples in curatedPCaData}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  warning = FALSE,
  message = FALSE,
  cache = TRUE
)
```

```{r packages}
library(curatedPCaData)
library(MultiAssayExperiment)
library(RaggedExperiment)
library(survival)
library(survminer)
library(corrplot)
```

# Example analyses with curatedPCaData

## Brief overview to MultiAssayExperiment functions

A handy guide for use of MAE objects is provided in: https://bioconductor.org/packages/release/bioc/vignettes/MultiAssayExperiment/inst/doc/QuickStartMultiAssay.html

A quick cheat sheet for interacting with MAE objects: https://bioconductor.org/packages/release/bioc/vignettes/MultiAssayExperiment/inst/doc/MultiAssayExperiment_cheatsheet.pdf

## Survival analysis

### Kaplan-Meiers

#### Gleason and biochemical recurrence

Traditional Gleason grade is the sum of Gleason major and Gleason minor. This can be exemplified as being clearly associated with biochemical recurrence (BCR) via lonk-rank tests in Kaplan-Meier curves:

```{r gleason_tcga, dpi=150, fig.width=7, fig.height=7, out.width="100%", fig.cap="Gleason grade group vs BCR (TCGA)"}
clin_tcga <- as.data.frame(colData(mae_tcga))
bcr_tcga <- Surv(time=clin_tcga$days_to_disease_specific_recurrence, event=clin_tcga$disease_specific_recurrence_status)

fit_gleason_tcga <- survfit(bcr_tcga ~ clin_tcga$grade_group)
ggsurvplot(fit_gleason_tcga, risk.table = TRUE, data = clin_tcga, pval=TRUE, title="TCGA BCR vs Gleason grade groups", xscale = "d_y", break.time.by = 365.25*2, xlim=365.25*c(0,10), xlab="Time (years)")
```

```{r gleason_taylor, dpi=150, fig.width=7, fig.height=7, out.width="100%", fig.cap="Gleason grade group vs BCR (Taylor et al.)"}
clin_taylor <- as.data.frame(colData(mae_taylor))
bcr_taylor <- Surv(time=clin_taylor$days_to_disease_specific_recurrence, event=clin_taylor$disease_specific_recurrence_status)

fit_gleason_taylor <- survfit(bcr_taylor ~ clin_taylor$grade_group)
ggsurvplot(fit_gleason_taylor, risk.table = TRUE, data = clin_taylor, pval=TRUE, title="Taylor et al. BCR vs Gleason grade groups", xscale = "d_y",, break.time.by = 365.25*2, xlim=365.25*c(0,10), xlab="Time (years)")
```

```{r gleason_barwick, dpi=150, fig.width=7, fig.height=7, out.width="100%", fig.cap="Gleason score vs BCR (Barwick et al.)"}
clin_barwick <- as.data.frame(as.matrix(colData(mae_barwick)))
bcr_barwick <- Surv(time=as.numeric(clin_barwick$days_to_disease_specific_recurrence), event=as.integer(clin_barwick$disease_specific_recurrence_status))

fit_gleason_barwick <- survfit(bcr_barwick ~ clin_barwick$gleason_grade)
ggsurvplot(fit_gleason_barwick, risk.table = TRUE, data = clin_barwick, pval=TRUE, title="Barwick et al. BCR vs Gleason grade groups")
```

#### Gleason and survival

One must consider the distribution of events when conducting survival modelling. One dataset which has enough end-points to model overall survival is Abida et al., which depicted enough end-points for death, while other datasets represented a more benign form of the disease and thus survival modelling for death is not really feasible.

```{r gleason_abida_death, dpi=150, fig.width=7, fig.height=7, out.width="100%", fig.cap="Gleason grade vs OS (Abida et al.)"}
clin_abida <- as.data.frame(as.matrix(colData(mae_abida)))
os_abida <- Surv(time=as.numeric(clin_abida$days_to_overall_survival), event=as.integer(clin_abida$overall_survival_status))

fit_gleason_abida <- survfit(os_abida ~ clin_abida$gleason_grade)
ggsurvplot(fit_gleason_abida, risk.table = TRUE, data = clin_abida, pval=TRUE, title="Abida et al. OS vs Gleason grade", xscale = "d_y", break.time.by = 365.25, xlim=365.25*c(0,6), xlab="Time (years)")
```

Rather than Gleason grade in the metastatic disease, the site of metastasis maybe of more interest:

```{r metassite_abida_death, dpi=150, fig.width=7, fig.height=7, out.width="100%", fig.cap="Metastasis site vs OS (Abida et al.)"}
fit_metassite_abida <- survfit(os_abida ~ clin_abida$metastatic_site)
ggsurvplot(fit_metassite_abida, risk.table = TRUE, data = clin_abida, pval=TRUE, title="Abida et al. OS vs Metastasis site", xscale = "d_y", break.time.by = 365.25, xlim=365.25*c(0,6), xlab="Time (years)")
```

Further, TMPRSS2-ERG fusion status (determined using gene expression) may be of interest:

```{r erg_abida_death, dpi=150, fig.width=7, fig.height=7, out.width="100%", fig.cap="Metastasis site vs OS (Abida et al.)"}
fit_erg_abida <- survfit(os_abida ~ clin_abida$ERG_fusion_GEX)
ggsurvplot(fit_erg_abida, risk.table = TRUE, data = clin_abida, pval=TRUE, title="Abida et al. OS vs ERG-fusion", xscale = "d_y", break.time.by = 365.25, xlim=365.25*c(0,6), xlab="Time (years)")
```

#### Genomic risk score


Sun et al. reports recurrence statuses (YES/NO, i.e. 1/0), but no follow-up times. We can still inspect a tabular representation of reported recurrence vs. transcriptome-calculated risk scores:

```{r gleason_sun_risks}
table(Recurrence = colData(mae_sun)$disease_specific_recurrence_status, DecipherAboveMedian = mae_sun[["scores"]]["decipher",] > median(mae_sun[["scores"]]["decipher",]))
table(Recurrence = colData(mae_sun)$disease_specific_recurrence_status, OncotypeAboveMedian = mae_sun[["scores"]]["oncotype",] > median(mae_sun[["scores"]]["oncotype",]))
table(Recurrence = colData(mae_sun)$disease_specific_recurrence_status, ProlarisAboveMedian = mae_sun[["scores"]]["prolaris",] > median(mae_sun[["scores"]]["prolaris",]))
```

## Cox proportional hazard models

### Decipher vs. BCR

```{r tcga_decipher_bcr}
common <- intersect(colnames(mae_tcga[["scores"]]), colData(mae_tcga)[,"sample_name"])
clin_tcga_subset <- clin_tcga[match(common, clin_tcga$sample_name),]
bcr_tcga_subset <- Surv(time=clin_tcga_subset$days_to_disease_specific_recurrence, event=clin_tcga_subset$disease_specific_recurrence_status)
tcga_decipher <- mae_tcga[["scores"]]["decipher",clin_tcga_subset[,"sample_name"]]
coxph_tcga_decipher <- coxph(bcr_tcga_subset ~ tcga_decipher)
summary(coxph_tcga_decipher)
```

```{r taylor_decipher_bcr}
common <- intersect(colnames(mae_taylor[["scores"]]), colData(mae_taylor)[,"sample_name"])
clin_taylor_subset <- clin_taylor[match(common, clin_taylor$sample_name),]
bcr_taylor_subset <- Surv(time=clin_taylor_subset$days_to_disease_specific_recurrence, event=clin_taylor_subset$disease_specific_recurrence_status)
taylor_decipher <- mae_taylor[["scores"]]["decipher",clin_taylor_subset[,"sample_name"]]
coxph_taylor_decipher <- coxph(bcr_taylor_subset ~ taylor_decipher)
summary(coxph_taylor_decipher)
```

### Oncotype DX vs. BCR

```{r tcga_oncotype_bcr}
tcga_oncotype <- mae_tcga[["scores"]]["oncotype",clin_tcga_subset[,"sample_name"]]
coxph_tcga_oncotype <- coxph(bcr_tcga_subset ~ tcga_oncotype)
summary(coxph_tcga_oncotype)
```

```{r taylor_oncotype_bcr}
taylor_oncotype <- mae_taylor[["scores"]]["oncotype",clin_taylor_subset[,"sample_name"]]
coxph_taylor_oncotype <- coxph(bcr_taylor_subset ~ taylor_oncotype)
summary(coxph_taylor_oncotype)
```

### Prolaris vs. BCR

```{r tcga_prolaris_bcr}
tcga_prolaris <- mae_tcga[["scores"]]["prolaris",clin_tcga_subset[,"sample_name"]]
coxph_tcga_prolaris <- coxph(bcr_tcga_subset ~ tcga_prolaris)
summary(coxph_tcga_prolaris)
```

```{r taylor_prolaris_bcr}
taylor_prolaris <- mae_taylor[["scores"]]["prolaris",clin_taylor_subset[,"sample_name"]]
coxph_taylor_prolaris <- coxph(bcr_taylor_subset ~ taylor_prolaris)
summary(coxph_taylor_prolaris)
```

## Correlations between genes and derived variables

### Risk scores, and AR score

```{r purity_risks_ar, dpi=150, fig.width=7, fig.height=7, out.width="100%", fig.cap="Correlations risk scores and AR score in TCGA"}
correlations <- cor(as.matrix(wideFormat(mae_tcga[,,"scores"])[,2:5]), use="na.or.complete")
corrplot::corrplot(correlations)
```

### TP53 and immune cell infiltration

```{r tp53mut_and_cibersort}

tp53 <- wideFormat(mae_tcga["TP53",,"mut"])



```

### Derived variables vs. end-points 

```{r tcellinf_bcr}



```

## Multi-omics

### Copy number deletion and gene expression

We choose SPOP as a concept example for mutation and copy-number alteration (i.e. deletion) tied to an endpoint (BCR) as well as as an example to exemplify connections between copy number alteration and transcriptome expression.

```{r tcga_spop}
spop_tcga <- wideFormat(mae_tcga["SPOP",,c("gex.fpkm", "cna.gistic", "mut")])
spop_tcga <- as.data.frame(spop_tcga)
```

Subsetting to just primary tumor samples:

```{r tcga_primary}
table(colData(mae_tcga)$sample_type)
spop_tcga <- spop_tcga[spop_tcga[,"primary"] %in% rownames(colData(mae_tcga)[which(colData(mae_tcga)$sample_type == "primary"),]),]
```

For example, it may be worth inspecting if the GISTIC2 discretized values {-2,-1,0,1,2} corresponding to deep deletion, shallow deletion, no somatic copy number change, low gain, and high gain, seem to correlate with the overall gene expression level:

```{r tcga_spop_gex_cna, dpi=150, fig.width=7, fig.height=7, out.width="100%", fig.cap="SPOP GEX vs CNA in TCGA"}
boxplot(spop_tcga[,"gex.fpkm_SPOP"] ~ spop_tcga[,"cna.gistic_SPOP"], main="TCGA CNA vs. GEX in SPOP", xlab="GISTIC2", ylab="Gene expression (FPKM)")
stripchart(spop_tcga[,"gex.fpkm_SPOP"] ~ spop_tcga[,"cna.gistic_SPOP"], pch=16, add=TRUE, vertical=TRUE)
```

As expected, samples with deletions called by GISTIC2 exhibit lower expression in general, while samples with copy number gains called by GISTIC2 have overall higher SPOP expression levels.

Let us pick an another gene of interest, PTEN. We can examine this also in other datasets; as an example, Ren et al with an N=65 overlap between GEX and CNA samples:

```{r ren_pten}
pten_ren <- wideFormat(mae_ren["PTEN",,c("gex.relz", "cna.gistic", "mut")])
pten_ren <- as.data.frame(pten_ren)
```

We notice that all samples in Ren et al. are primary samples; this is because they are presented already in comparison to a panel of normals:

```{r ren_primary}
table(colData(mae_ren)$sample_type)
```

```{r ren_pten_gex_cna, dpi=150, fig.width=7, fig.height=7, out.width="100%", fig.cap="PTEN GEX vs CNA in Ren et al."}
boxplot(pten_ren[,"gex.relz_PTEN"] ~ pten_ren[,"cna.gistic_PTEN"], main="Ren et al. CNA vs. GEX in PTEN", xlab="GISTIC2", ylab="Gene expression (relative z-scores)")
stripchart(pten_ren[,"gex.relz_PTEN"] ~ pten_ren[,"cna.gistic_PTEN"], pch=16, add=TRUE, vertical=TRUE)
```

We notice that the tumor for which the GISTIC2 call indicated a shallow or a deep deletion tended to have a lower expression of PTEN. Therefore, it makes sense to consider cross-omics information, when this is available.


### Multivariable regression from multiple assay types

Let us build a multivariable Cox regression model, which takes into account multiple omics as well as a couple clinical variables:

```{r tcga_multivar}
subset_tcga <- wideFormat(mae_tcga[c("SPOP", "PTEN", "TP53"),,c("gex.fpkm", "cna.gistic")], colDataCols=c("psa", "ERG_fusion_GEX", "disease_specific_recurrence_status", "days_to_disease_specific_recurrence", "sample_type"))
subset_tcga <- as.data.frame(subset_tcga)
subset_tcga <- subset_tcga[subset_tcga$sample_type == "primary",]

x <- subset_tcga[,c("psa", "ERG_fusion_GEX", "gex.fpkm_SPOP", "gex.fpkm_PTEN", "gex.fpkm_TP53", "cna.gistic_SPOP", "cna.gistic_PTEN", "cna.gistic_TP53")]
x <- as.matrix(x)

y <- Surv(time=subset_tcga$days_to_disease_specific_recurrence, event=subset_tcga$disease_specific_recurrence_status)

coxph_example_tcga <- coxph(y ~ x)
summary(coxph_example_tcga)

```

### Oncoprints

curatedPCaData offers wrapper functions to collapse over omics for creating oncoprints. ```ComplexHeatmap```-package offers convenient oncoprint functionality, which will be exemplified here. Let us define a subset of interesting genes and then query a MAE object with both gene expression and mutation data to collapse them into an oncoprint:

```{r oncoprintify, dpi=150, fig.width=7, fig.height=7, out.width="100%", fig.cap="Taylor et al. oncoprint with any gene and sample hits visualized"}
library(ComplexHeatmap)

# Genes of interest
genes <- c("TP53", "PTEN", "FGFR3", "BRCA2")

# Define how the alterations are shown in the oncoprint
alter_fun = list(
    background = function(x, y, w, h) 
        grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = "#CCCCCC", col = NA)),
    # Amplifications
    "Gain" = function(x, y, w, h) 
        grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = "orange", col = NA)),
    "High gain" = function(x, y, w, h) 
        grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = "red", col = NA)),
    # Deletions
    "Deep deletion" = function(x, y, w, h) 
        grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = "blue", col = NA)),
    "Deletion" = function(x, y, w, h) 
        grid.rect(x, y, w*0.9, h*0.9, gp = gpar(fill = "cyan", col = NA)),
    # Missense SNV
    "Missense_Mutation" = function(x, y, w, h) 
        grid.rect(x, y, w*0.7, h*0.7, gp = gpar(fill = "green", col = NA))
)

# Wrapper function helps sweep over omics in a MAE object
taylor_oncoprint_1 <- curatedPCaData:::wrapper_oncoprintify(mae_taylor, genes=genes)

# Draw the oncoprint
ComplexHeatmap::oncoPrint(taylor_oncoprint_1, alter_fun = alter_fun)

```

However, it should be noted that the overlap between omics is not perfect. Therefore, if one wishes to print an oncoprint with only fully present data, parameter ```join = "both"``` should be used:

```{r oncoprintify2, dpi=150, fig.width=7, fig.height=7, out.width="100%", fig.cap="Taylor et al. oncoprint with only intersection of genes and samples available on both omics visualized"}

# Only samples with both CNA and mutation calls and for which all genes are present
taylor_oncoprint_2 <- curatedPCaData:::wrapper_oncoprintify(mae_taylor, genes=genes, join = "both")

# Draw the oncoprint
ComplexHeatmap::oncoPrint(taylor_oncoprint_2, alter_fun = alter_fun)


```

### General trends in gene expression across datasets

Using the curatedpcadata package, patterns in gene expression across different datasets can be studied.

```{r correlation_plot}

calculate_correlation<-function(dataset,gene){
  name=deparse(substitute(dataset))
  data_subset=as.data.frame(t(dataset[gene,]))
  
  data_subset$sample=rownames(data_subset)
  data_subset[,1]=as.numeric(data_subset[,1])
  
  dataset=as.data.frame(t(dataset))
  
  
  cor_values=numeric(0)
  
  for( i in 1:ncol(dataset)){
    indvidual_cor_values=print(cor(dataset[,i],data_subset[,1],method = "spearman"))
    cor_values <- c(cor_values, indvidual_cor_values) 
  }
  
  cor_matrix=data.frame(matrix(NA, nrow = ncol(dataset), ncol = 2))
  colnames=colnames(dataset)[1:ncol(dataset)]
  colnames=as.vector(colnames)
  cor_matrix[,1]=colnames
  cor_matrix[,2]=cor_values
  #paste0("spearman_correlation_coeff_",deparse(substitute(dataset)))
  colnames(cor_matrix)=c("gene",name)
  return(cor_matrix)
}


correlation_of_correlations<-function(gene){
  tcga=as.data.frame(curatedPCaData::mae_tcga[["gex.fpkm"]])
  taylor=as.data.frame(curatedPCaData::mae_taylor[["gex.rma"]])
  abida=as.data.frame(curatedPCaData::mae_abida[["gex.relz"]])
  ren=as.data.frame(curatedPCaData::mae_ren[["gex.relz"]])
  barbieri=as.data.frame(curatedPCaData::mae_barbieri[["gex.relz"]])
  igc=as.data.frame(curatedPCaData::mae_igc[["gex.rma"]])
  friedrich=as.data.frame(curatedPCaData::mae_friedrich[["gex.logq"]])
  barwick=as.data.frame(curatedPCaData::mae_barwick[["gex.logq"]])
  chandran=as.data.frame(curatedPCaData::mae_chandran[["gex.rma"]])
  icgcca=as.data.frame(curatedPCaData::mae_icgcca[["gex.rma"]])
  kim=as.data.frame(curatedPCaData::mae_kim[["gex.rma"]])
  kunderfranco=as.data.frame(curatedPCaData::mae_kunderfranco[["gex.logr"]])
  sun=as.data.frame(curatedPCaData::mae_sun[["gex.rma"]])
  true=as.data.frame(curatedPCaData::mae_true[["gex.logr"]])
  wallace=as.data.frame(curatedPCaData::mae_wallace[["gex.rma"]])
  wang=as.data.frame(curatedPCaData::mae_wang[["gex.rma"]])
  weiner=as.data.frame(curatedPCaData::mae_weiner[["gex.rma"]])
  
  cor_matrix_tcga = calculate_correlation(tcga,gene)
  cor_matrix_taylor= calculate_correlation(taylor,gene)
  cor_matrix_abida= calculate_correlation(abida,gene)
  cor_matrix_ren= calculate_correlation(ren,gene)
  cor_matrix_barbieri= calculate_correlation(barbieri,gene)
  cor_matrix_igc= calculate_correlation(igc,gene)
  cor_matrix_friedrich= calculate_correlation(friedrich,gene)
  cor_matrix_barwick= calculate_correlation(barwick,gene)
  cor_matrix_chandran= calculate_correlation(chandran,gene)
  cor_matrix_icgcca= calculate_correlation(icgcca,gene)
  cor_matrix_kim= calculate_correlation(kim,gene)
  cor_matrix_kunderfranco= calculate_correlation(kunderfranco,gene)
  cor_matrix_sun= calculate_correlation(sun,gene)
  cor_matrix_true= calculate_correlation(true,gene)
  cor_matrix_wallace= calculate_correlation(wallace,gene)
  cor_matrix_wang= calculate_correlation(wang,gene)
  cor_matrix_weiner= calculate_correlation(weiner,gene)
  
  
  
  #put all data frames into list
  df_list <- list(cor_matrix_tcga, cor_matrix_taylor, cor_matrix_abida,cor_matrix_ren,
                  cor_matrix_barbieri,cor_matrix_igc,cor_matrix_friedrich,cor_matrix_barwick,
                  cor_matrix_chandran,cor_matrix_icgcca,cor_matrix_kim,
                  cor_matrix_kunderfranco,cor_matrix_sun,cor_matrix_true,cor_matrix_wallace,
                  cor_matrix_wang,cor_matrix_weiner)
  
  combined_cor_list=Reduce(function(x, y) merge(x, y, all=F), df_list)  
  rownames(combined_cor_list)=combined_cor_list[,"gene"]
  combined_cor_list=combined_cor_list[,-1]
  
  combined_cor_list[,c(1:17)]=sapply( combined_cor_list[,c(1:17)], as.numeric )
  return(combined_cor_list)
}

```

For example the correlation of AR expression to the expression of all the other genes across all datasets can be studied with the help of a correlation plot as shown:

```{r, results=FALSE}

library(corrplot)
combined_cor_list=correlation_of_correlations("AR")
corr_of_corr = cor(combined_cor_list)
```

```{r, dpi=150, fig.width=7, fig.height=7, out.width="100%"}
corrplot::corrplot(corr_of_corr,method = 'color')

```

# Session info

```{r session}
sessionInfo()
```

